Return empty store tree on non-existing version

RiccardoM opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description
This PR tries to fix the different store versions loading problem. It simply returns an empty tree when a store version does not exist. Closes: #7385.

This has been tested using the following procedure:

1. Start a Desmos chain **without** the `relationships` module. 
2. Create an upgrade proposal that switches to a new Desmos version **with** the new `relationships` module registered
3. Vote that upgrade proposal, and let it run. 
4. Try querying the `relationships`. 

Everything seems fine. The new module can be queried without any problem, as well as the old ones.

### TODO
If this change is OK, the following things still have to be done: 

- [x] Change the `store #GetImmutable` documentation accordingly
- [x] Fix the unit tests properly
- [x] Add a CHANGELOG entry


---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

alexanderbez commented:
Nice! So how does this work under the following scenarios:

- Querying an existing store at some previous non-pruned height H
- Querying an existing store at some previous pruned height H<sub>p</sub>
- Querying an existing store at current height H<sub>c</sub>
- Querying a new store at some previous non-pruned height H
- Querying a new store at some previous pruned height H<sub>p</sub>
- Querying a new store at current height H<sub>c</sub>

/cc @erikgrinaker

erikgrinaker commented:
Yeah, I don't really see how this can be right. If the app is at height 900 and the new store is at version 100 (such that version 1-100 in the new store map to heights 800-900 in the app) then surely querying height 850 for the new store will not work since there is no version 850 in the store.

Since you say that this works, then maybe the SDK ends up querying the latest version in this case, but the query requested height 850 not the latest height so it'll still be wrong.

RiccardoM commented:
> Yeah, I don't really see how this can be right. If the app is at height 900 and the new store is at version 100 (such that version 1-100 in the new store map to heights 800-900 in the app) then surely querying height 850 for the new store will not work since there is no version 850 in the store.
> 
> Since you say that this works, then maybe the SDK ends up querying the latest version in this case (I'm not too familiar with the code), but the query requested height 850 not the latest height so it'll still be wrong.

I think in this case, since we're returning an empty tree, then all the queries will just be empty (which is correct since the store didn't exist). Am I missing something?

erikgrinaker commented:
>I think in this case, since we're returning an empty tree when the version does not exist, then all the queries will just be empty (which is correct since the store didn't exist). Am I missing something?

I think so, because we're returning an empty store for the wrong heights.

In the example above, version 1-100 in the new store was created during heights 800-900. If I write "foo=850" at height 850, this will be stored at version 50 in the new store.

If I query `foo` as of height 850 it will return nothing, because the store has no version 850 - this is wrong, since the value at height 850 should be `foo=850`.

If I query `foo` as of height 50 it will return `foo=850`. This is wrong, since the store did not exist at all at height 50, and should return empty.

RiccardoM commented:
> In the example above, version 1-100 in the new store was created during heights 800-900. If I write "foo=850" at height 850, this will be stored at version 50 in the new store.
> 
> If I query `foo` as of height 850 it will return nothing, because the store has no version 850 - this is wrong, since the value at height 850 should be `foo=850`.
> 
> If I query `foo` as of height 50 it will return `foo=850`. This is wrong, since the store did not exist at all at height 50, and should return empty.

That makes sense. So, is there any way we can get the height at which the store was created? Or, maybe, we can set the new store height with the upgrade height and pad all the previous versions with empty stores? Could this be possible?

erikgrinaker commented:
Yeah, so I don't want to prescribe a solution for the SDK here, since I'm not on that team - I'll defer to @alexanderbez. But in IAVL 0.15.0-rc2 we implemented an initial version API which allows setting the initial version for a new IAVL store, via `Options.InitialVersion` and/or `MutableTree.SetInitialVersion()`. I have backported this to 0.14 on this branch, but not cut a release yet: https://github.com/cosmos/iavl/tree/0.14.x

It's definitely a viable solution to create a new store with a specific initial version, and then ignore missing version errors when attempting to fetch any previous versions (although this can have the effect of silently ignoring actual bugs that cause versions to go missing). It's also possible to pad out the store with empty versions, corresponding to the versions that exist in the other stores (which will depend on pruning settings).

alexanderbez commented:
> It's definitely a viable solution to create a new store with a specific initial version, and then ignore missing version errors when attempting to fetch any previous versions (although this can have the effect of silently ignoring actual bugs that cause versions to go missing).

This is the route I'd like to attempt to go with.

1. Set initial versions for _new_ stores on upgrades. Maybe we can automatically do this in the root multi-store during `LoadVersion` so app devs don't have to worry about this during app construction. And,
2. Ignore the store when the version is not available.

RiccardoM commented:
> > It's definitely a viable solution to create a new store with a specific initial version, and then ignore missing version errors when attempting to fetch any previous versions (although this can have the effect of silently ignoring actual bugs that cause versions to go missing).
> 
> This is the route I'd like to attempt to go with.
> 
> 1. Set initial versions for _new_ stores on upgrades. Maybe we can automatically do this in the root multi-store during `LoadVersion` so app devs don't have to worry about this during app construction. And,
> 2. Ignore the store when the version is not available.

Do you have any hint how this can be achieved? Even high level ones, I'll try to do this tomorrow

alexanderbez commented:
Use the `SetInitialVersion` API. I think when loading stores, we can tell if a store is new (i.e. latest version is zero). If so, we can call `SetInitialVersion` on it. Would that work @erikgrinaker?

erikgrinaker commented:
I think so, yeah.

RiccardoM commented:
@alexanderbez @erikgrinaker Ok, I've tried implementing it as you said. I had to introduce a little hack to make it work properly: when using `tree.LoadVersion`, I ignore any error if the store has been added recently. This is to avoid the following error when using `tree.SetInitialVersion`:

```
initial version set to 2, but found earlier version 1
```

Apparently, when loading a specific version of the tree, the IAVL checks to make sure that the initial version of each store is not greater than any other previous versions of any store.

I've added a test set name `TestLoadStore` inside `store_test.go`, but I was not able to find a way to prune a specific height to test such cases.

erikgrinaker commented:
>initial version set to 2, but found earlier version 1

Doesn't this indicate that the fix doesn't work? It's too late to set the initial version if the tree already has a version in it.

I don't think you can call `Version()` until the tree has been loaded, it always returns `0` in that case. You may have to call `Load()` first, which will load the latest available version, but that may not be optimal here since we're given a specific version to load.

RiccardoM commented:
> > initial version set to 2, but found earlier version 1
> 
> Doesn't this indicate that the fix doesn't work? It's too late to set the initial version if the tree already has a version in it.
> 
> I don't think you can call `Version()` until the tree has been loaded, it always returns `0` in that case. You may have to call `Load()` first, which will load the latest available version, but that may not be optimal here since we're given a specific version to load.

Ok, I've tried moving the `SetInitialVersion` after loading the store. I still have some doubts about the tests. Are they correct? Cause I don't think so to be honest

alexanderbez commented:
I believe the version should be set before loading the store. Somewhere in the root multi-store when we load all the versions on bootstrapping, we need to detect that a store is _new_ (i.e. its version is zero) and the last commit id version is higher, so we call SetInitialVersion on that store when creating the IAVL store.

RiccardoM commented:
> I believe the version should be set before loading the store. Somewhere in the root multi-store when we load all the versions on bootstrapping, we need to detect that a store is _new_ (i.e. its version is zero) and the last commit id version is higher, so we call SetInitialVersion on that store when creating the IAVL store.

Can you please point me to where this should happen then? I'm not familiar with the multi store but I get dig into this having a starting place where to look at (method name).

alexanderbez commented:
`app.LoadLatestVersion` -> `BaseApp.storeLoader` -> `store.LoadLatestVersion` -> `store.getLatestVersion` & `store.loadVersion` -> `store.loadCommitStoreFromParams` -> `iavl.LoadStore`.


So most likely we'll need to slightly modify `loadCommitStoreFromParams` and maybe even `iavl.LoadStore`.

RiccardoM commented:
When taking a look at the `rootmulti/store.go` file, I noticed there is a method called `LoadLatestVersionAndUpgrade`. I checked it at it's used inside `upgrade/types/UpgradeStoreLoader`. 

I then checked the `store.StoreUpgrades` types and noticed how this is thought to represent the upgrades to the store that should be done when an automatic upgrade is performed. 

So here's what I thought of doing: 

1. I edited the `StoreUpgrades` adding a new `Added` field.  
   This represents the modules that should be added when loading the store after the upgrade.

2. Inside `rootmulti.store.go`, I edited the `loadVersion` method.  
   If a store key is present inside the `StoreUpgrades.Added` field, its initial version is set inside the `storeParams.initialVersion` field that I have added.

3. Inside `iavl/store.go` I added a new method called `LoadStoreWithInitialVersion`.  
   This simply uses `NewMutableTreeWithOpts` to set the initial version before loading the tree. It is called only from `loadVersion` when a store `initialVersion` is not 0. 

You can take a look at this on my latest commit. Can something like this work @alexanderbez @erikgrinaker? 

I tried changing the `LoadLatestVersionAndUpgrade` tests to include this, but I'm not sure if the tests I've written are correct.

alexanderbez commented on go.mod:
```diff
@@ -4,6 +4,7 @@ module github.com/cosmos/cosmos-sdk
 
 require (
 	github.com/99designs/keyring v1.1.6
+	github.com/DataDog/zstd v1.4.5 // indirect
```
Why are there go.mod change?

RiccardoM replied:
I just updated Tendermint dependency to point to the branch that was created having the fixes, since no release was created. Then running `go mod tidy` the `go.mod` file changed.

alexanderbez commented on store/iavl/store.go:
```diff
@@ -80,9 +87,10 @@ func UnsafeNewStore(tree *iavl.MutableTree) *Store {
 func (st *Store) GetImmutable(version int64) (*Store, error) {
 	if !st.VersionExists(version) {
-		return nil, iavl.ErrVersionDoesNotExist
+		return &Store{tree: &immutableTree{&iavl.ImmutableTree{}}}, nil
```
Remind me why we're doing this?

RiccardoM replied:
It was added to make sure that quering a store using a height before the upgrade (the time in which it was created) doesn't return an error. Can we remove this now or is it still necessary?

alexanderbez replied:
As long as we ensure an error isn't returned and the query is handled gracefully 👍

RiccardoM replied:
Ok I've tried reverted this and the `TestMultistoreLoadWithUpgrade` test is still passing so I think we can keep it as it was.

RiccardoM replied:
Nevermind, this is still necessary otherwise some queries will fail

alexanderbez commented on store/iavl/store.go:
```diff
@@ -43,7 +43,14 @@ type Store struct {
+// LoadStore returns an IAVL Store as a CommitKVStore setting its initialVersion
+// to the one given. Internally, it will load the store's version (id) from the
+// provided DB. An error is returned if the version fails to load.
+func LoadStoreWithInitialVersion(db dbm.DB, id types.CommitID, lazyLoading bool, initialVersion int64) (types.CommitKVStore, error) {
```
We should take a `uint64` here instead of converting/casting, which can panic.

RiccardoM replied:
I've removed this and placed the casting inside `loadVersion`. Unfortunately, we have to perform at least once cast since the `ver` parameter there has type `int64` while the `LoadStoreWithInitialVersion` method now accepts a `uint64`.

alexanderbez reviewed (approved):
utACK

erikgrinaker commented on store/rootmulti/store.go:
```diff
@@ -813,7 +820,15 @@ func (rs *Store) loadCommitStoreFromParams(key types.StoreKey, id types.CommitID
+		var store types.CommitKVStore
+		var err error
+
+		if params.initialVersion == 0 {
+			store, err = iavl.LoadStore(db, id, rs.lazyLoading)
+		} else {
+			store, err = iavl.LoadStoreWithInitialVersion(db, id, rs.lazyLoading, params.initialVersion)
+		}
```
This check isn't strictly necessary, since `LoadStore()` is equivalent to `LoadStoreWithInitialVersion(..., 0)`. But might be good to be explicit about it anyway.

erikgrinaker commented on store/iavl/store.go:
```diff
@@ -43,7 +43,14 @@ type Store struct {
+// LoadStore returns an IAVL Store as a CommitKVStore setting its initialVersion
+// to the one given. Internally, it will load the store's version (id) from the
+// provided DB. An error is returned if the version fails to load.
+func LoadStoreWithInitialVersion(db dbm.DB, id types.CommitID, lazyLoading bool, initialVersion uint64) (types.CommitKVStore, error) {
```
Note that this will probably have to be modified once IAVL 0.15 is released. In 0.14, loading a missing version from an _empty_ tree (no versions) does not error, which has been changed/fixed in 0.15: https://github.com/cosmos/iavl/pull/317

alexanderbez replied:
Thanks for the update. We'll just have to update this once that version is released.

erikgrinaker reviewed (approved):
Well done! Let me know when we're ready to cut 0.14.1, i.e. when we're certain we won't need any changes to the `InitialVersion` functionality.

amaurymartiny reviewed (approved):
utACK!

There are 3 todo items in the original comment, @RiccardoM would you like to address them before we merge this in?

RiccardoM commented:
@erikgrinaker I was trying to backport this against v0.39.1 to test it with a local Desmos instance to make sure the error does not show up. Unluckily, I think that tendermint 0.33.8 is not compatible and I get the following errors: 

```
# github.com/tendermint/tendermint/state/txindex/kv
../../Go/pkg/mod/github.com/tendermint/tendermint@v0.33.8/state/txindex/kv/kv.go:143:76: undefined: db.SetDeleter
# github.com/tendermint/tendermint/lite/proxy
../../Go/pkg/mod/github.com/tendermint/tendermint@v0.33.8/lite/proxy/verifier.go:25:60: multiple-value db.NewDB() in single-value context
# github.com/tendermint/tendermint/consensus
../../Go/pkg/mod/github.com/tendermint/tendermint@v0.33.8/consensus/replay_file.go:280:15: assignment mismatch: 1 variable but db.NewDB returns 2 values
../../Go/pkg/mod/github.com/tendermint/tendermint@v0.33.8/consensus/replay_file.go:284:10: assignment mismatch: 1 variable but db.NewDB returns 2 values
make: *** [Makefile:23: build] Error 2
```

Is there a way to port such changes into a branch so that I can build Cosmos 0.39.1 using your IAVL branch properly?

erikgrinaker commented:
Hi @RiccardoM. It looks like you are using the wrong versions of dependencies - those errors are caused by tm-db 0.6.0, but both IAVL 0.14, Cosmos SDK 0.39.1, and Tendermint 0.33.8 depend on tm-db 0.5.1. Note that Cosmos SDK `master` depends on IAVL 0.15, tm-db 0.6, and Tendermint 0.34.0-rcX.

RiccardoM commented:
@alexanderbez @erikgrinaker I've tested this porting the changes to a local branch based on Cosmos v0.39.1 and, unluckily, this does **not** solve the problem. 

I've execute the following actions:

1. run a chain without a module;
2. create a proposal to upgrade to a new version adding a new module;
3. perform some queries. 

Point (1) and (2) work properly. However, point 3 has a problem. 

If we keep the `GetImmutable` as it is right not, `iavl.ErrVersionDoesNotExist` is returned. 

So, I tried making it return an immutable store backed by an empty tree. This makes the queries work properly for old stores. However, when trying to query a new store it will always return empty results. I've tested this by doing the following.

1. After the upgrade, perform a transaction using the new module. In this case I added a relationship. 
2. Once the transaction is successful, try to query the data specifying:
   1. a height *previous* to the update
   2. the latest height

What happens is that both (i) and (ii) always return empty results, even though the transaction was successful. 

Do you by any chance have any idea on this?

alexanderbez commented:
> So, I tried making it return an immutable store backed by an empty tree. This makes the queries work properly for old stores. However, when trying to query a new store it will always return empty results. 

Right, so when the new store is loaded, are you sure it's loaded at the same version as the rest of the stores? i.e. the versions are not out of sync?

If so, I would add debugging logs to when the store(s) are loaded during a query `CacheMultiStoreWithVersion`. Add logs to `GetImmutable` also. See if the version exists. If it doesn't, then that's the problem. If it does, then when can dig from there.

RiccardoM commented:
> Right, so when the new store is loaded, are you sure it's loaded at the same version as the rest of the stores? i.e. the versions are not out of sync?
> 
> If so, I would add debugging logs to when the store(s) are loaded during a query `CacheMultiStoreWithVersion`. Add logs to `GetImmutable` also. See if the version exists. If it doesn't, then that's the problem. If it does, then when can dig from there.

Of, so I've added a few logs and this are the outputs.

### During the upgrade
```
------------- Key: KVStoreKey{0xc000395bc0, gov} -------------
Loading store
Version: 49 
Params initial version: 0

Loaded tree
Version: 49
Initial version: 0

Last commit id: 49

------------- Key: KVStoreKey{0xc000395c40, relationships} -------------
Loading store
Version: 0 
Params initial version: 49

Loaded tree
Version: 0
Initial version: 49

Last commit id: 0
```

### After the upgrade, when querying some data
```
------------- Key: KVStoreKey{0xc000395c20, profiles} -------------
Getting immutable store
Version: 67
Required version: 67

------------- Key: KVStoreKey{0xc000395c40, relationships} -------------
Getting immutable store
Version: 66
Required version: 67
```

### After stopping the chain at height 74 and trying to start it again
```
------------- Key: KVStoreKey{0xc000fdfcf0, relationships} -------------
Loading store
Version: 73 
Params initial version: 74
failed to load Store: initial version set to 74, but found earlier version 49
```

### Conclusion
It looks like that after the upgrade, the store has a version which is exactly `H - 1` where `H` is the version of all the other stores. 

Any idea how to fix this?

erikgrinaker commented:
Sounds like an off-by-one error in the initial version. If all the stores are currently at version H, then the initial version of the new store must be H+1, since that will be the next version that is saved.

RiccardoM commented:
> Sounds like an off-by-one error in the initial version. If all the stores are currently at version H, then the initial version of the new store must be H+1, since that will be the next version that is saved (and thus the initial version to be saved for the new store).

Thanks! That was it! I'm going to do some more tests, fix the code, clean the documentation and then this PR should be ready.

RiccardoM commented:
> Well done! Let me know when we're ready to cut 0.14.1, i.e. when we're certain we won't need any changes to the `InitialVersion` functionality.

Can you please cut it? After the tests I've done, there are no other changes needed

erikgrinaker commented:
>Can you please cut it? After the tests I've done, there are no other changes needed

Sure. I'm heading into a meeting shortly, will cut a release after.

erikgrinaker commented:
It occurred to me that another option would be to simply set the initial version to H and then explicitly save an empty version at the upgrade height. It might be more intuitive that the store exists from the upgrade height onwards (inclusive), but don't know if this would complicate the implementation. Don't have any strong opinions on this though, just an idea.

RiccardoM commented:
> It occurred to me that another option would be to simply set the initial version to H and then explicitly save an empty version at the upgrade height. It might be more intuitive that the store exists from the upgrade height onwards (inclusive), but don't know if this would complicate the implementation. Don't have any strong opinions on this though, just an idea.

@alexanderbez Do you have any opinion on this? Should we leave it as it is setting the initial value to `H + 1`, or should we make it `H` and save an empty commit?

alexanderbez commented:
I don't see any immediate benefit to having an empty commit. I think the H+1 approach should be fine. Glad we figured it out.

erikgrinaker commented:
@RiccardoM IAVL 0.14.1 is out: https://github.com/cosmos/iavl/releases/tag/v0.14.1

RiccardoM commented:
@erikgrinaker @alexanderbez @amaurymartiny I think this is now ready to be merged. I'm also working on another branch to port this to v0.39.1 as well

alexanderbez commented on store/types/store.go:
```diff
@@ -63,6 +64,19 @@ type StoreRename struct {
+	if s == nil {
+		return false
+	}
+	for _, d := range s.Added {
+		if d == key {
+			return true
+		}
+	}
+	return false
```
```suggestion
	return tmstrings.StringInSlice(s.Added, key)
```

RiccardoM replied:
Neat!

alexanderbez reviewed (changes requested):
one minor nit, otherwise LGTM 👍

erikgrinaker commented:
I've released IAVL 0.15.0-rc4, which now errors when attempting to load a non-existing version from an empty tree. This change may affect this PR. See: https://github.com/cosmos/iavl/pull/317
